### PR TITLE
MDRange Tile Size Tuning

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
Implements #3156 

If a user doesn't specify a tile size, this PR will add tuning hooks to enable that tuning.

Calling out a few oddities

- Removal of an include [here](https://github.com/kokkos/kokkos/compare/develop...DavidPoliakoff:feature/mdrange_tuning?expand=1#diff-d24f8e17b98a0cb7576e3f28ad7dec4cd2ce4542ece9415e546a27d7e7b9b75eL54). This was necessary to avoid some loop of includes
- Changing some of the infrastructure [here](https://github.com/kokkos/kokkos/compare/develop...DavidPoliakoff:feature/mdrange_tuning?expand=1#diff-d24f8e17b98a0cb7576e3f28ad7dec4cd2ce4542ece9415e546a27d7e7b9b75eL363). It turns out we changed the interface of `init`, which I missed.
- This weird [generic_tune_policy thing](https://github.com/kokkos/kokkos/compare/develop...DavidPoliakoff:feature/mdrange_tuning?expand=1#diff-18a17cd6a96df955c44fa267dc82f323d397e432e34c9f40807b44ff595110f4R279). Honestly this is a bit premature. But I've noticed that both policy tuning things I wanted to do had the same rough "shape" so I'll try to do it generically.

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
Blocked on my ability to understand this error: https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/3941/pipeline

**Comment by crtrott** [comment]
Looks like you are trying to invoke a rank 3 thing with only two indicies somewhere?

**Comment by DavidPoliakoff** [comment]
How so? The MDRange being pointed out is [2d](https://github.com/kokkos/kokkos/blob/master/core/unit_test/incremental/Test14_MDRangeReduce.hpp#L125), no?

**Comment by DavidPoliakoff** [comment]
@masterleinad : https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/4043/pipeline/43 the error persist after merging latest `develop`

**Comment by masterleinad** [comment]
Hmmm... I guess you don't have a configuration at hand that would reproduce the error?

**Comment by DavidPoliakoff** [comment]
I can build one. Is it about seeing the skipped contexts?

**Comment by masterleinad** [comment]
> I can build one. Is it about seeing the skipped contexts?

No, the changes here are small enough that it should not be too hard to figure out what exactly is breaking it by selectively disabling changes.

**Comment by DavidPoliakoff** [comment]
Also, since I made it, here's a build with `-ftemplate-backtrace-limit=0`

```
/ascldap/users/dzpolia/src/kokkos/core/src/impl/KokkosExp_IterateTileGPU.hpp(72): error: function "lambda [](int, int, Test::value_type &)->void" cannot be called with the given argument list
            argument types are: (const int, const int)
            object type is: const lambda [](int, int, Test::value_type &)->void
          detected during:
            instantiation of "std::enable_if_t<std::is_void<Tag>::value, void> Kokkos::Impl::_tag_invoke<Tag,Functor,Args...>(const Functor &, Args &&...) [with Tag=void, Functor=lambda [](int, int, Test::value_type &)->void, Args=<const int &, const int &>]"
(144): here
            instantiation of "void Kokkos::Impl::DeviceIterateTile<2, PolicyType, Functor, Tag>::exec_range() const [with PolicyType=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Functor=lambda [](int, int, Test::value_type &)->void, Tag=void]"
/ascldap/users/dzpolia/src/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel.hpp(559): here
            instantiation of "void Kokkos::Impl::ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>, Kokkos::Cuda>::operator()() const [with FunctorType=lambda [](int, int, Test::value_type &)->void, Traits=<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>]"
/ascldap/users/dzpolia/src/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp(121): here
            instantiation of "void Kokkos::Impl::cuda_parallel_launch_local_memory(DriverType) [with DriverType=Kokkos::Impl::ParallelFor<lambda [](int, int, Test::value_type &)->void, Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::CudaSpace::execution_space>]"
/ascldap/users/dzpolia/src/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp(322): here
            instantiation of "std::decay_t<decltype((<expression>))> Kokkos::Impl::CudaParallelLaunchKernelFunc<DriverType, Kokkos::LaunchBounds<0U, 0U>, Kokkos::Impl::Experimental::CudaLaunchMechanism::LocalMemory>::get_kernel_func() [with DriverType=Kokkos::Impl::ParallelFor<lambda [](int, int, Test::value_type &)->void, Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::CudaSpace::execution_space>]"
/ascldap/users/dzpolia/src/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp(653): here
            [ 6 instantiation contexts not shown ]
            instantiation of "void Kokkos::Tools::Impl::ReductionSwitcher<Kokkos::InvalidType>::tune(size_t, const std::__cxx11::string &, ExecPolicy &, const Functor &, const TagType &) [with Functor=lambda [](int, int, Test::value_type &)->void, TagType=Kokkos::ParallelReduceTag, ExecPolicy=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>]"
/ascldap/users/dzpolia/src/kokkos/core/src/impl/Kokkos_Profiling.hpp(529): here
            instantiation of "void Kokkos::Tools::Impl::begin_parallel_reduce<ReducerType,ExecPolicy,FunctorType>(ExecPolicy &, FunctorType &, const std::__cxx11::string &, uint64_t &) [with ReducerType=Kokkos::InvalidType, ExecPolicy=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, FunctorType=const lambda [](int, int, Test::value_type &)->void]"
/ascldap/users/dzpolia/src/kokkos/core/src/Kokkos_Parallel_Reduce.hpp(862): here
            instantiation of "void Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(const std::__cxx11::string &, const PolicyType &, const FunctorType &, ReturnType &) [with PolicyType=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, FunctorType=lambda [](int, int, Test::value_type &)->void, ReturnType=Test::value_type]"
/ascldap/users/dzpolia/src/kokkos/core/src/Kokkos_Parallel_Reduce.hpp(1018): here
            instantiation of "std::enable_if<Kokkos::is_execution_policy<PolicyType>::value, void>::type Kokkos::parallel_reduce(const PolicyType &, const FunctorType &, ReturnType &) [with PolicyType=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, FunctorType=lambda [](int, int, Test::value_type &)->void, ReturnType=Test::value_type]"
/ascldap/users/dzpolia/src/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp(128): here
            instantiation of "void Test::TestMDRangeReduce<ExecSpace>::reduce_MDRange() [with ExecSpace=Kokkos::Cuda]"
/ascldap/users/dzpolia/src/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp(178): here

/ascldap/users/dzpolia/src/kokkos/core/src/impl/KokkosExp_IterateTileGPU.hpp(72): error: function "lambda [](int, int, Test::value_type &)->void" cannot be called with the given argument list
            argument types are: (const int, const int)
            object type is: const lambda [](int, int, Test::value_type &)->void
          detected during:
            instantiation of "std::enable_if_t<std::is_void<Tag>::value, void> Kokkos::Impl::_tag_invoke<Tag,Functor,Args...>(const Functor &, Args &&...) [with Tag=void, Functor=lambda [](int, int, Test::value_type &)->void, Args=<const int &, const int &>]"
(144): here
            instantiation of "void Kokkos::Impl::DeviceIterateTile<2, PolicyType, Functor, Tag>::exec_range() const [with PolicyType=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Functor=lambda [](int, int, Test::value_type &)->void, Tag=void]"
/ascldap/users/dzpolia/src/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel.hpp(559): here
            instantiation of "void Kokkos::Impl::ParallelFor<FunctorType, Kokkos::MDRangePolicy<Traits...>, Kokkos::Cuda>::operator()() const [with FunctorType=lambda [](int, int, Test::value_type &)->void, Traits=<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>]"
/ascldap/users/dzpolia/src/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp(121): here
            instantiation of "void Kokkos::Impl::cuda_parallel_launch_local_memory(DriverType) [with DriverType=Kokkos::Impl::ParallelFor<lambda [](int, int, Test::value_type &)->void, Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::CudaSpace::execution_space>]"
/ascldap/users/dzpolia/src/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp(322): here
            instantiation of "std::decay_t<decltype((<expression>))> Kokkos::Impl::CudaParallelLaunchKernelFunc<DriverType, Kokkos::LaunchBounds<0U, 0U>, Kokkos::Impl::Experimental::CudaLaunchMechanism::LocalMemory>::get_kernel_func() [with DriverType=Kokkos::Impl::ParallelFor<lambda [](int, int, Test::value_type &)->void, Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, Kokkos::CudaSpace::execution_space>]"
/ascldap/users/dzpolia/src/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp(653): here
            [ 6 instantiation contexts not shown ]
            instantiation of "void Kokkos::Tools::Impl::ReductionSwitcher<Kokkos::InvalidType>::tune(size_t, const std::__cxx11::string &, ExecPolicy &, const Functor &, const TagType &) [with Functor=lambda [](int, int, Test::value_type &)->void, TagType=Kokkos::ParallelReduceTag, ExecPolicy=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>]"
/ascldap/users/dzpolia/src/kokkos/core/src/impl/Kokkos_Profiling.hpp(529): here
            instantiation of "void Kokkos::Tools::Impl::begin_parallel_reduce<ReducerType,ExecPolicy,FunctorType>(ExecPolicy &, FunctorType &, const std::__cxx11::string &, uint64_t &) [with ReducerType=Kokkos::InvalidType, ExecPolicy=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, FunctorType=const lambda [](int, int, Test::value_type &)->void]"
/ascldap/users/dzpolia/src/kokkos/core/src/Kokkos_Parallel_Reduce.hpp(862): here
            instantiation of "void Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(const std::__cxx11::string &, const PolicyType &, const FunctorType &, ReturnType &) [with PolicyType=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, FunctorType=lambda [](int, int, Test::value_type &)->void, ReturnType=Kokkos::View<Test::value_type, Kokkos::Cuda>]"
/ascldap/users/dzpolia/src/kokkos/core/src/Kokkos_Parallel_Reduce.hpp(1018): here
            instantiation of "std::enable_if<Kokkos::is_execution_policy<PolicyType>::value, void>::type Kokkos::parallel_reduce(const PolicyType &, const FunctorType &, ReturnType &) [with PolicyType=Kokkos::MDRangePolicy<Kokkos::Cuda, Kokkos::Rank<2U, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>, FunctorType=lambda [](int, int, Test::value_type &)->void, ReturnType=Kokkos::View<Test::value_type, Kokkos::Cuda>]"
/ascldap/users/dzpolia/src/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp(136): here
            instantiation of "void Test::TestMDRangeReduce<ExecSpace>::reduce_MDRange() [with ExecSpace=Kokkos::Cuda]"
/ascldap/users/dzpolia/src/kokkos/core/unit_test/incremental/Test14_MDRangeReduce.hpp(178): here
```

**Comment by masterleinad** [comment]
Hmmm.. now `OpenMPTarget` is complaining:
```
/var/jenkins/workspace/Kokkos/core/src/Kokkos_Tuners.hpp:533:30: error: no matching member function for call to 'get_max_mdrange_tile_size'
    int max_tile_size = calc.get_max_mdrange_tile_size(policy, functor, tag);
                        ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_Profiling.hpp:339:24: note: in instantiation of function template specialization 'Kokkos::Tools::Experimental::MDRangeTuner<5>::MDRangeTuner<Test::(anonymous namespace)::TestMDRange_5D<Kokkos::Serial>, Kokkos::ParallelReduceTag, Kokkos::Tools::Impl::Impl::ComplexReducerSizeCalculator<Kokkos::Sum<double>>, Kokkos::Serial, Kokkos::Rank<5, Kokkos::Iterate::Default, Kokkos::Iterate::Default>, Kokkos::IndexType<int>>' requested here
                       Tuner(label, policy, functor, tag,
```

**Comment by DavidPoliakoff** [comment]
I can fix it, I probably forgot an overload

**Comment by DavidPoliakoff** [comment]
Yup, I forgot to change another form of bounds calculator to match the new interface, fixed

File: core/src/Kokkos_Tuners.hpp
Comment (masterleinad) [inline core/src/Kokkos_Tuners.hpp:503]:
```suggestion
    fill_tile(cont[x], tile_size / x);
```

File: core/src/Kokkos_Tuners.hpp
Comment (masterleinad) [inline core/src/Kokkos_Tuners.hpp:526]: Why not obey the rule of zero if we can.
```suggestion
```

**Review by masterleinad — COMMENTED** [review]
Just a few comments. Otherwise, the changes look reasonable to me.

File: core/src/Kokkos_Tuners.hpp
Comment (DavidPoliakoff) [reply core/src/Kokkos_Tuners.hpp]: Much more elegant, good catch

File: core/src/Kokkos_Tuners.hpp
Comment (DavidPoliakoff) [reply core/src/Kokkos_Tuners.hpp]: Good call

**Comment by DavidPoliakoff** [comment]
@masterleinad : done, thanks, Worked on my build, will see how CI treats it

**Comment by masterleinad** [comment]
You need to fix indentation.

**Comment by DavidPoliakoff** [comment]
This apparently needs another round of fixes to adjust for #3626

**Comment by masterleinad** [comment]
> This apparently needs another round of fixes to adjust for #3626

Sorry :confused:

**Review by masterleinad — APPROVED** [review]

File: core/src/impl/Kokkos_Profiling.hpp
Comment (dalg24) [inline core/src/impl/Kokkos_Profiling.hpp:364]: Document these overloads (single line above each of them would do).  I had to stare at their signature wondering what was different.

File: core/src/impl/Kokkos_Profiling.hpp
Comment (dalg24) [inline core/src/impl/Kokkos_Profiling.hpp:242]: `get_mdrange_max_tile_size_product`?

**Review by dalg24 — DISMISSED** [review]
Please post a use example in the discussion

File: core/src/Kokkos_Tuners.hpp
Comment (DavidPoliakoff) [reply core/src/Kokkos_Tuners.hpp:509]: I thought `constexpr const` meant something different than just `const`? Am I misunderstanding, or did that change? (Nothing changes as much as the meaning of const)

**Comment by DavidPoliakoff** [comment]
@dalg24, what do you mean by

```
Please post a use example in the discussion
```


? Any MDRangePolicy kernel without a specified tile size will exercise this

File: core/src/Kokkos_Tuners.hpp
Comment (masterleinad) [reply core/src/Kokkos_Tuners.hpp:509]: `constexpr` member functions are not necessarily `const` but variables are.

File: core/src/Kokkos_Tuners.hpp
Comment (DavidPoliakoff) [reply core/src/Kokkos_Tuners.hpp]: Removed in 065fead

File: core/src/impl/Kokkos_Profiling.hpp
Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_Profiling.hpp]: Implemented in 065fead

File: core/src/Kokkos_Tuners.hpp
Comment (DavidPoliakoff) [reply core/src/Kokkos_Tuners.hpp]: Regardless, compiles, so implemented in 065fead

File: core/src/impl/Kokkos_Profiling.hpp
Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_Profiling.hpp:372]: 065fead

File: core/src/impl/Kokkos_Profiling.hpp
Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_Profiling.hpp]: Sure, 065fead

**Review by jrmadsen — APPROVED** [review]
Do we have something like:

```cpp
#if CXXSTD >= 17
# define KOKKOS_IF_CONSTEXPR if constexpr
#else
# define KOKKOS_IF_CONSTEXPR if
#endif
```
?

Because a few things like: `if(should_tune(policy))` appear to be compile-time constants and in that case, we should use `if constexpr` when available

**Comment by dalg24** [comment]
Please cleanup history and I will merge

**Comment by DavidPoliakoff** [comment]
Hey @jrmadsen , those actually won't be compile_time, but it's a good thought if we can. "should_tune" just checks that a given instance of a policy should be tuned, which depends on runtime parameters (in the TeamPolicy case, was either parameter declared "AUTO". In the MDRange case, was the tile size unspecified?)

@dalg24, will do, thanks

**Comment by DavidPoliakoff** [comment]
@dalg24 cool, passed CI, this is ready to merge in my book

**Pull request merged**
